#  Looking for most profitable apps by analysing Android & iOS  Apps market

    
This project will be analysing the profiles of dataset of
mobile phone apps that are currently available on both the 
Apple Store and Google Play markets.
     
The aim is to use the conclussion of the anaysis to predict
and decide on developing future apps that will be profitable on 
both platforms.

# Opening and Exploring relevant dataset

The datasets to be used can be described as follows:

1. Android apps dataset containing approximately 10,000 Android apps from Google Play; This can be viewed and downloaded from
     [GooglePlay](https://www.kaggle.com/lava18/google-play-store-apps#googleplaystore.csv)

2. Apple ios dataset containing about 7,000 iOS apps from the App Store is availbale from this link- [Applestore](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).

In the code below, we will:    

* open the dataset file with `open('Applestore.csv')` command and assigned to variable `opened_file`.
* import `reader` command from `csv` module
* pass the opened file to be read by the `reader(opened_file)` command and assigned to variable `read_file`
* transform the opened and read file into list of lists by passing into `list(read_file)` command and result assigned to variable `apps_data`


In [1]:
from csv import reader

opened_file =  open('AppleStore.csv', encoding="utf-8")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

opened_file = open('googleplaystore.csv', encoding="utf-8")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]


A function named `explore_data()` is created and coded below to be used to explore the 2 data sets: `AppleStore.csv` and `googleplaystore.csv`. The function has the buil-in option tha can be used to explore the number of rows and columns of any data set passed into it                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            .

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = ios[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
       

In [3]:
print(ios_header)
print('\n')
explore_data(ios,1,3,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


In [4]:
print(android_header)
print('\n')
explore_data(android,0,3,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 10841
Number of columns: 13


Exploring these datasets, the following columns will be required for further investigation and required for the anaysis to be done:
* For AppleStore: "Price", "prime_genre", "Cont_rating", "user_rating"
* For GooglePlay: "rating", "Price", "Genre", Content_rating", "Type"

The complete list and descriptions of the columns for both dataset
can be seen here for [Applestore](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) and for [GooglePlay](https://www.kaggle.com/lava18/google-play-store-apps#googleplaystore.csv)

# Removing wrong data
From the discussion forum, the index for the row that has the error is 10472. As it is not certain whether this row index takes account of the header being included or not.

Further investigation of the dataset with the header shows the error at 10472.

The coding and results below investigates further and show the result


In [5]:
# checking for the actual index of the row with error
for row in android:
    if len(row) != len(android_header):
        print(row)
        #print('\n')
        print(android.index(row))
        print('\n')   

# checking the length of the original dataset against length after
print(len(android))
del android[10472]
print(len(android))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


10841
10840


# Exploring the dataset to remove duplicates entries
Exploring the Google play data set, several apps have duplicate entries; for example "Instagram" has 4, "8 Ball Pool"-8, "Baby Panda Care"- 2, "MTV- 2.
The last two being straight forward duplicates to be removed

# Part One

In [6]:
for row_entry in android:
    name = row_entry[0]
    if (name == 'Instagram') or (name == 'MTV'):
        print(row_entry,'\n')

['MTV', 'ENTERTAINMENT', '3.8', '35279', '15M', '1,000,000+', 'Free', '0', 'Teen', 'Entertainment', 'July 8, 2018', '11.45.0', '4.4 and up'] 

['MTV', 'ENTERTAINMENT', '3.8', '35279', '15M', '1,000,000+', 'Free', '0', 'Teen', 'Entertainment', 'July 8, 2018', '11.45.0', '4.4 and up'] 

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 

['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device'] 



In [7]:
duplicate_entry = []
unique_entry = []

for row_entry in android:
    name = row_entry[0]
    if name in unique_entry:
        duplicate_entry.append(name)
    else:
        unique_entry.append(name)
        
print('Number of duplicated rows in the data set:', len(duplicate_entry)) 
print('\n')
print('Some examples of duplicated row entry:', duplicate_entry[:11])

Number of duplicated rows in the data set: 1181


Some examples of duplicated row entry: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic']


# Removing duplicate entries - Part Two
Analysis of data set requires that duplicate data need to be taken care of one way or the other.
Exploring each case of the duplicated row entry for  an app, the duplicate entry will highlight the differences and hence the criterion necessary for resolution.



In [8]:
reviews_max = {}
for row_entry in android:
    name = row_entry[0]
    n_reviews = float(row_entry[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
        #print(reviews_max)
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Inspecting and checking the length of `reviews_max` dictionary and compare with actual length of original data set; and noting the length of duplicates above

In [9]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


Dictionary `reviews_max` with the key `name` and value `n_reviews`created and used ealier is to to be used in the coding below to remove the duplicate data entry rows

In [10]:
android_clean = []
already_added = []
for row_entry in android:
    name = row_entry[0]
    n_reviews = float(row_entry[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(row_entry)
        already_added.append(name)
   




The `explore_data()` function is being used below to explore the veracity of the `android_clean` data set to confirm the correct numbers of row and coulumns expected

In [11]:
explore_data(android_clean, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 9659
Number of columns: 13


# Section to remove Non-English apps - Part-1
The data sets are checked below for non-English characters in some of the apps name . 

In [12]:
print(ios[813][1])
print(ios[6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])
print('\n')


爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ




# Section to remove Non-English apps - Part-2
A fucntion is being created and is used to explore and test each row of the data set to check if the app name is English or not.
The feature of ASCI charcters of being betwenn  0 and 127 and the built-in function `ord()` is used to test and isolate non-English apps.

In [13]:
def is_app_english(string):
    non_asci = 0
    
    for character in string:
        if ord(character) > 127:
            non_asci += 1
        
    if non_asci > 3:
        return False
    else:
        return True

print(is_app_english('Docs To Go™ Free Office Suite'))
print(is_app_english('Instachat 😜'))
print(is_app_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))


True
True
False


The `is_app_english()` is now to be used on each data set to romove non-english apps. The process is to create initial two separete empty lists `android_english[]` and `ios_english[]` to which apps identified as English are appended or added. 

In [14]:
# Separating English and non_english apps from Android data set

android_english = []
android_non_english = []

for row_entry in android_clean:
    name = row_entry[0]
    if is_app_english(name): 
        android_english.append(row_entry)
    else:
        android_non_english.append(name)
        

 
print('Number of Android English apps:', len(android_english))
print('\n')
print('Number of Android non-English apps:', len(android_non_english))
print('\n')
print('Some examples of Android non-English apps:', android_non_english[:3])

explore_data(android_english, 0, 3, True)



Number of Android English apps: 9614


Number of Android non-English apps: 45


Some examples of Android non-English apps: ['Flame - درب عقلك يوميا', 'သိင်္ Astrology - Min Thein Kha BayDin', 'РИА Новости']
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 9614
Number of columns: 13


The ios data set is now being innvestigated below and as commented.

In [15]:
# checking for the actual index of the row with error
for row in ios:
    if len(row) != len(ios_header):
        print(row)
        #print('\n')
        print(ios.index(row))
        print('\n')   

# checking the length of the dataset 

print(len(ios))

7197


In [16]:
# checking for duplicate entries in ios data set:

ios_duplicate_entry = []
ios_unique_entry = []

for row_entry in ios:
    name = row_entry[1]
    if name in ios_unique_entry:
        ios_duplicate_entry.append(name)
    else:
        ios_unique_entry.append(name)
        
print('Number of duplicated rows in ios data set:', len(ios_duplicate_entry)) 
print('\n')
print('Some examples of duplicated row entry:', ios_duplicate_entry[:11])

Number of duplicated rows in ios data set: 2


Some examples of duplicated row entry: ['Mannequin Challenge', 'VR Roller Coaster']


Duplicate data need to be romoved from ios data set
Exploring each case of the duplicated row entry for  an app, the duplicate entry will highlight the differences and hence the criterion necessary for resolution.

In [17]:
# Separating English and non_english apps from ios data set

ios_english = []
ios_non_english = []

for row_entry in ios:
    name = row_entry[1]
    if is_app_english(name): 
        ios_english.append(row_entry)
    else:
        ios_non_english.append(row_entry)
        
print('Number of ios English apps:', len(ios_english)) 
print('\n')
print('Number of ios non-English apps:', len(ios_non_english))
print('\n')
print('Some examples of Android non-English apps:', ios_non_english[:3])

Number of ios English apps: 6183


Number of ios non-English apps: 1014


Some examples of Android non-English apps: [['445375097', '爱奇艺PPS -《欢乐颂2》电视剧热播', '224617472', 'USD', '0.0', '14844', '0', '4.0', '0.0', '6.3.3', '17+', 'Entertainment', '38', '5', '3', '1'], ['405667771', '聚力视频HD-人民的名义,跨界歌王全网热播', '90725376', 'USD', '0.0', '7446', '8', '4.0', '4.5', '5.0.8', '12+', 'Entertainment', '24', '4', '1', '1'], ['336141475', '优酷视频', '204959744', 'USD', '0.0', '4885', '0', '3.5', '0.0', '6.7.0', '12+', 'Entertainment', '38', '0', '2', '1']]


In [18]:
# Explorung the ios data set to confirm the numbers of column and rows

explore_data(ios_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


## Isolating Free apps from each data set

In [19]:
# Separating Free and Non_Free apps from Android data set

Free_android_apps = []
Non_Free_android_apps = []

for row_entry in android_english:
    price = row_entry[7]
    if price == '0':
        Free_android_apps.append(row_entry)
    else:
        Non_Free_android_apps.append(row_entry)
        
print('Number of  Free Android apps:', len(Free_android_apps)) 
print('\n')
print('Number of Non Free Android apps:', len(Non_Free_android_apps))
print('\n')
print('Some examples of Free Android apps:', Free_android_apps[:3])
print('\n')
print('Some examples of Non Free Android apps:', Non_Free_android_apps[:3])

Number of  Free Android apps: 8864


Number of Non Free Android apps: 750


Some examples of Free Android apps: [['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


Some examples of Non Free Android apps: [['TurboScan: scan documents and receipts in PDF', 'BUSINESS', '4.7', '11442', '6.8M', '100,000+', 'Paid', '$4.99', 'Everyone', 'Business', 'March 25, 2018', '1.5.2', '4.0 and up'], ['Tiny Scanner Pro: PDF Doc Scan', 'BUSINESS', '4.8', '10295', '39M', '100,000+', 'Paid', '$4.99', 'Everyon

In [20]:
# Exploring the Free Android apps dataset as follows:

explore_data(Free_android_apps, 0, 3, True)
    

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 8864
Number of columns: 13


In [21]:
# Separating Free and Non_Free apps from ios data set

Free_ios_apps = []
Non_Free_ios_apps = []

for row_entry in ios_english:
    name = row_entry[1]
    price = row_entry[4]
    if price == '0.0': 
        Free_ios_apps.append(row_entry)
    else:
        Non_Free_ios_apps.append(row_entry)
        
print('Number of  Free ios apps:', len(Free_ios_apps)) 
print('\n')
print('Number of Non Free ios apps:', len(Non_Free_ios_apps))
print('\n')
print('Some examples of Free ios apps:', Free_ios_apps[:3])
print('\n')
print('Some examples of Non Free ios apps:', Non_Free_ios_apps[:3])

Number of  Free ios apps: 3222


Number of Non Free ios apps: 2961


Some examples of Free ios apps: [['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']]


Some examples of Non Free ios apps: [['362949845', 'Fruit Ninja Classic', '104590336', 'USD', '1.99', '698516', '132', '4.5', '4.0', '2.3.9', '4+', 'Games', '38', '5', '13', '1'], ['500116670', 'Clear Vision (17+)', '37879808', 'USD', '0.99', '541693', '69225', '4.5', '4.5', '1.1.3', '17+', 'Games', '43', '5', '1', '1'], ['479516143', 'Minecraft: Pocket Edition', '147787776', 'USD', '6.99', '522012', '1148', '4.5', '4.5', '1.1', '9+', 'Games', '37', '1', '11', '1']]

In [22]:
# Summarising and exploring the Free ios apps dataset as follows:

explore_data(Free_ios_apps, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 3222
Number of columns: 16


In [23]:
# ongoing analysis and exploring each data set shows the following
print('Number of Free ios apps:', len(Free_ios_apps))
print('Number of Free Android apps:', len(Free_android_apps))


Number of Free ios apps: 3222
Number of Free Android apps: 8864


# Looking for most popular apps by genre by generating frequency tables
The stage is now to continue the analysis by looking into 
each data set to discover and study the profiles of what 
makes the most popular apps.
The aim of the analysis being to develop free to downloand and install apps with the objective of being able to generate revenue and profits through advertisements.
The requirement then becomes the need for an app that has a profile of most popular genre with largest number of users or downloads.

The criterion of each data set to study could therefore be amongst others: genres, number of downloads, ratings, number of reviews etc.


For Apple Store dataset, the columns of interest are
    * trac_name(app name,)
    * rating_count_tot
    * content_rating
    * prime_genre
    
 For Google Play, the columns are
    * App(app name)
    * category
    * Installs(downloads)
    * Genres

column `prime_genre` at index 11 of the Apple Store data set is to be used in the `extrac()` function below which is in turn to be used in the `freq_table()` to get the most popular or common genres. 

In [24]:
def extract(index):
    column = []    
    for row in Free_ios_apps[1:]:
        value = row[index]
        column.append(value)    
    return column

ios_genres = extract(11)

def extract(index):
    column = []    
    for row in Free_android_apps[1:]:
        value = row[index]
        column.append(value)    
    return column

android_genres = extract(9)


def freq_table(a_list):
    table = {}
    for item in a_list:
        if item in table:
            table[item] += 1
        else:
            table[item] = 1            
    return table
    
ios_genres_ft = freq_table(ios_genres)
android_genres_ft = freq_table(android_genres)

print('Frequenct Table for free ios apps:')
print(ios_genres_ft)
print('\n')
print('There are',len(ios_genres_ft), 'types of free ios genres as above')
print('\n')
print('Frequency Table for free Android apps:')
print(android_genres_ft)
print('\n')
print('There are',len(android_genres_ft), 'types of free android apps as above')


Frequenct Table for free ios apps:
{'Photo & Video': 160, 'Games': 1874, 'Music': 66, 'Social Networking': 105, 'Reference': 18, 'Health & Fitness': 65, 'Weather': 28, 'Utilities': 81, 'Travel': 40, 'Shopping': 84, 'News': 43, 'Navigation': 6, 'Lifestyle': 51, 'Entertainment': 254, 'Food & Drink': 26, 'Sports': 69, 'Book': 14, 'Finance': 36, 'Education': 118, 'Productivity': 56, 'Business': 17, 'Catalogs': 4, 'Medical': 6}


There are 23 types of free ios genres as above


Frequency Table for free Android apps:
{'Art & Design': 52, 'Art & Design;Creativity': 6, 'Auto & Vehicles': 82, 'Beauty': 53, 'Books & Reference': 190, 'Business': 407, 'Comics': 54, 'Comics;Creativity': 1, 'Communication': 287, 'Dating': 165, 'Education': 474, 'Education;Creativity': 4, 'Education;Education': 30, 'Education;Pretend Play': 5, 'Education;Brain Games': 3, 'Entertainment': 538, 'Entertainment;Brain Games': 7, 'Entertainment;Creativity': 3, 'Entertainment;Music & Video': 15, 'Events': 63, 'Finance': 328

The frequency table is now to be used below in combination with percentage calculations for each relevant columns of `genre`, `prime_genre' and 'category'.

In [25]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        item = row[index]
        if item in table:
            table[item] += 1
        else:
            table[item] = 1            
            
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total)*100
        table_percentages[key] = percentage
        
    return table_percentages
    

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
    

In [26]:
display_table(Free_ios_apps, 11) # Prime Genre

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


# Analysing the frequency table for the `prime_genre` column of Apple Store data
The following are found out:
* The most common genre is Games at 58% and the runner-up is
   Entertainment at around 7%.
* There is a pattern of more than half of the apps being of Games and
   Entertainment.
* The general impression therefore is that most popular or numerous
  apps are for Entertainment such as Games,Photo & video, social
  networking, sports and music and less popular apps for practical
  purposes like education, shopping, utilities, productivity,
  lifestyle.
* On the basis of current analysis, an app with profiles similar to  
  those of Games apps can be recommended; although othr factors will
  have to be further considered.
* However, there is need to investigate further whether large numbers
  for an app genre will translate to large number of users.

In [27]:
display_table(Free_android_apps, -4) # Genres from Google Play store

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [28]:
display_table(Free_android_apps, 1) # Category from Google Play store

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

**Analysing the frequency table for `Category` & `Genres` columns of Google play data set**
The following are found out:
* The most common genre is Tools(8%)  and Family(18%) in `Genre` and
   `Category` columns respectively. 
* There is a pattern of a large and disparate clasification into
   genres in the `Genres` column with the highest share not up to
   10%. And for the `Category` column, there is a pattern of apps
   at the top of the table that Family or Entertaimnet based.
   Generally, for the differnces between the items in *GENRE* and 
   *Category* are not distinct enough or not clear


***Compairing the patterns on both data set, Games apps are most popular on Apple Store at more than 55% while Tools  & Family apps are most popular on Google Store. And by compariosn, Games apps on Google Store is less than 10% ***



** Finding and displaying the most popular app genre in the
Apple Store data set

In [29]:

print('Average number of user ratings per app genre on the App Store:')

ios_genre_apps_ft = freq_table(Free_ios_apps, 11)

for genre in ios_genre_apps_ft:
    total = 0
    len_genre = 0
    for app in Free_ios_apps:
        genre_app = app[11]
        if genre_app == genre:
            n_user_ratings = float(app[5])
            total += n_user_ratings
            len_genre += 1
            
    avg_n_user_ratings = total / len_genre
    print(genre, ':', avg_n_user_ratings)
  

Average number of user ratings per app genre on the App Store:
Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


The top 3 Apple Store apps with the highest number of average ratings numbers are:
* Navigation - 86, 090
* Reference - 74, 942
* Social Networking - 71,548

Any of these top three can be the recommended app for development, but the
data for both Navigation and Social Networking are skewed because of the 
presence and contributions of data from big players like Waze, Google Maps, Facebook, pinterest, Skype etc. Also, the same argument will apply to Music with figures being skewed by big players like Spotify.

So Navigation, Social Networking and Music apps might not be as popular as the figures suggest.


In [30]:
for app in Free_ios_apps:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The disproportionate amount of figures for user ratings for Navigation can be 
seen above.

For *Reference* with average user rating of 74,942 the numbers are skewed by Bible and Dictionary.com figures as seen below.

In [31]:
for app in Free_ios_apps:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5]) # print app name and ratings number

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


The recommendation can therefore be narrowed down to a Reference-type app thereby avoiding the over-crowded market of fun apps like Games and Entertainment as well as avoiding competing with big names players like Google, Facebook, Spotify etc in apps where they already have unasailable leads

# Analysing for popular Apps by Genre 
# on Google Play.

  The equivalent relevant column in Google Play store is the "installs" column.


In [32]:
display_table(Free_android_apps, 5) # print the installs from Google Play store

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


As above information is not precise as to the actual number of installs, approximation are being applied to get some idea.
This require converting the sting-style install number format to float type and 
generate a frequency table.

Generating the frequency table for the process of calculating the
average number of intalls per app genre for the Google Play data set.


In [33]:

print('Average number of installs per app genre on Google Play data set:')

android_category_apps_ft = freq_table(Free_android_apps, 1)

for category in android_category_apps_ft:
    total = 0
    len_category = 0
    for app in Free_android_apps:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
       
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)


Average number of installs per app genre on Google Play data set:
ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHE

In [34]:
# As the COMMUNICATION app has the highest number of installs, we need to 
# investgate

print('List of COOMUNICATION-type app:')

for app in Free_android_apps:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

List of COOMUNICATION-type app:
WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & 

The figure for COMMUNICATION-type apps installed is skewed by few very large
number of installs like WhatsApp, Facebook Messenger with 1 billion+ installs
and others in the 500,000,000+.

We want to remove these high-number installs and calculate the average number of installs without them as follows:


In [35]:

# code to remove all the communication apps that have over 100 million installs,
# the average would be reduced roughly ten times:

under_100_m = []

for app in Free_android_apps:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

The pattern discovered for the COMMUNICATION-type app is also seen for apps
of types: VIDEO_PLAYERS, SOCIAL AND PHOTOGRAPHY dominated by few giants and will be difficult to compete with.

Similarlly the GAME-type app also seem popular with average installs of over 15million, but the market already appears to be saturated.

As we had breifly explored the REFERENCE-type app in Apple Store data, similar
type on Google play store id BOOK_REFERENCE  which seem popular with average 
installs of 8,767,811.

This type of app is explored further below:



In [36]:
for app in Free_android_apps:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

As before, there are a few apps belonging to few giants and whose number of
installs will skew the average numbers.
we can check for apps whose installs are 100,000,000+ as below:


In [37]:
for app in Free_android_apps:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


There are not many apps for this type and so there might be potential for app
of this type that have average installs of between 1,000,000+ and 
50,000,000+. This might then give some idea of app to be recommended

In [38]:
for app in Free_android_apps:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

The top item above appears to be niche apps of software for accessing and reading book,for libraries and dictionaries; hence competition will be encountered.

Data shows that popular book or reference works like Quran could be popular with additions of features like games, audion, video, daily quotes and notifications.

# Conclussion

For this project, data sets from both Apple Store nad Google Play were analyzed 
with the aim of finding and recommending an app that will be profitable on both platforms.

The conclussion and recommendation is for an app that will be based on a popular 
book; prefereably a  recent one. The app will then have to be developed with features for audio and video interactions for things like games, quizzes based on the book content and a forum for group discussions.